src/perf/lib/PerfHelpers.h

```diff
@@ -0,0 +1,629 @@
+                if (SecurityConfig == nullptr) {
+                    WriteOutput("Failed to create security configuration for thumbprint:'%s'.\n", certThumbprint);
+                    return QUIC_STATUS_INVALID_PARAMETER;
+                }
+            } else {
+                uint32_t machineCert = 0;
+                TryGetValue(argc, argv, "machine_cert", &machineCert);
+                QUIC_CERTIFICATE_HASH_STORE_FLAGS flags =
+                    machineCert ? QUIC_CERTIFICATE_HASH_STORE_FLAG_MACHINE_STORE : QUIC_CERTIFICATE_HASH_STORE_FLAG_NONE;
+
+                SecurityConfig = GetSecConfigForThumbprintAndStore(MsQuic, Registration, flags, certThumbprint, certStoreName);
+                if (SecurityConfig == nullptr) {
+                    WriteOutput(
+                        "Failed to create security configuration for thumbprint:'%s' and store: '%s'.\n",
+                        certThumbprint,
+                        certStoreName);
+                    return QUIC_STATUS_INVALID_PARAMETER;
+                }
+            }
+        }
+        return QUIC_STATUS_SUCCESS;
+    }
+
+    ~MsQuicSecurityConfig() {
+        if (SecurityConfig) {
+            MsQuic->SecConfigDelete(SecurityConfig);
+        }
+    }
+
+    operator QUIC_SEC_CONFIG*() const { return SecurityConfig; }
+
+    QUIC_SEC_CONFIG* SecurityConfig {nullptr};
+};
+
+struct ListenerScope {
+    HQUIC Handle;
+    ListenerScope() : Handle(nullptr) { }
+    ListenerScope(HQUIC handle) : Handle(handle) { }
+    ~ListenerScope() { if (Handle) { MsQuic->ListenerClose(Handle); } }
+    operator HQUIC() const { return Handle; }
+};
+
+struct ConnectionScope {
+    HQUIC Handle;
+    ConnectionScope() : Handle(nullptr) { }
+    ConnectionScope(HQUIC handle) : Handle(handle) { }
+    ~ConnectionScope() { if (Handle) { MsQuic->ConnectionClose(Handle); } }
+    operator HQUIC() const { return Handle; }
+};
+
+struct StreamScope {
+    HQUIC Handle;
+    StreamScope() : Handle(nullptr) { }
+    StreamScope(HQUIC handle) : Handle(handle) { }
+    ~StreamScope() { if (Handle) { MsQuic->StreamClose(Handle); } }
+    operator HQUIC() const { return Handle; }
+};
+
+struct EventScope {
+    QUIC_EVENT Handle;
+    EventScope() { QuicEventInitialize(&Handle, FALSE, FALSE); }
+    EventScope(QUIC_EVENT event) : Handle(event) { }
+    ~EventScope() { QuicEventUninitialize(Handle); }
+    operator QUIC_EVENT() const { return Handle; }
+};
+
+struct QuicBufferScope {
+    QUIC_BUFFER* Buffer;
+    QuicBufferScope() : Buffer(nullptr) { }
+    QuicBufferScope(uint32_t Size) : Buffer((QUIC_BUFFER*) new uint8_t[sizeof(QUIC_BUFFER) + Size]) {
+        QuicZeroMemory(Buffer, sizeof(*Buffer) + Size);
+        Buffer->Length = Size;
+        Buffer->Buffer = (uint8_t*)(Buffer + 1);
+    }
+    operator QUIC_BUFFER* () { return Buffer; }
+    ~QuicBufferScope() { if (Buffer) { delete[](uint8_t*) Buffer; } }
+};
+
+
+struct CountHelper {
```

Comment:
Looks like this is essentially a QUIC_RUNDOWN_REF, right?

Reply:
It was actually copied from QuicPing, as its way of counting connections. But needs a way to take an external event. Also the helper functions.